File: Java/CesiumLanguageWriterTests/translatedSrc/cesiumlanguagewritertests/TestJulianDateConversions.java
package cesiumlanguagewritertests;


import agi.foundation.compatibility.*;
import agi.foundation.compatibility.Action;
import agi.foundation.compatibility.ArgumentException;
import agi.foundation.compatibility.AssertHelper;
import agi.foundation.compatibility.DateTimeHelper;
import agi.foundation.compatibility.TestContextRule;
import agi.foundation.TypeLiteral;
import cesiumlanguagewriter.*;
import java.time.ZonedDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.Collection;
import javax.annotation.Nonnull;
import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.FixMethodOrder;
import org.junit.Rule;
import org.junit.runners.MethodSorters;
import org.junit.Test;

/**
 * A series of tests to exercise the various date and time types.
 */
@SuppressWarnings({
    "unused",
    "deprecation",
    "serial"
})
@FixMethodOrder(MethodSorters.NAME_ASCENDING)
public class TestJulianDateConversions {
    /**
    * Tests conversion from a {@link JulianDate} to a {@link ZonedDateTime}.
    */
    @Test
    public final void julianToDateTime() {
        JulianDate julianDate = new JulianDate(2451545.0, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        ZonedDateTime dateTime = julianDate.toDateTime();
        AssertHelper.assertEquals(2000, dateTime.getYear());
        AssertHelper.assertEquals(1, dateTime.getMonthValue());
        AssertHelper.assertEquals(1, dateTime.getDayOfMonth());
        AssertHelper.assertEquals(12, dateTime.getHour());
        AssertHelper.assertEquals(0, dateTime.getMinute());
        AssertHelper.assertEquals(0, dateTime.getSecond());
        AssertHelper.assertEquals(0, DateTimeHelper.getMillisecond(dateTime));
        AssertHelper.assertEquals(ZoneOffset.UTC, dateTime.getZone());
        julianDate = new JulianDate(2453736.5, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        dateTime = julianDate.toDateTime();
        AssertHelper.assertEquals(2006, dateTime.getYear());
        AssertHelper.assertEquals(1, dateTime.getMonthValue());
        AssertHelper.assertEquals(1, dateTime.getDayOfMonth());
        AssertHelper.assertEquals(0, dateTime.getHour());
        AssertHelper.assertEquals(0, dateTime.getMinute());
        AssertHelper.assertEquals(0, dateTime.getSecond());
        AssertHelper.assertEquals(0, DateTimeHelper.getMillisecond(dateTime));
        AssertHelper.assertEquals(ZoneOffset.UTC, dateTime.getZone());
        julianDate = new JulianDate(2441683.5, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        dateTime = julianDate.toDateTime();
        AssertHelper.assertEquals(1973, dateTime.getYear());
        AssertHelper.assertEquals(1, dateTime.getMonthValue());
        AssertHelper.assertEquals(1, dateTime.getDayOfMonth());
        AssertHelper.assertEquals(0, dateTime.getHour());
        AssertHelper.assertEquals(0, dateTime.getMinute());
        AssertHelper.assertEquals(0, dateTime.getSecond());
        AssertHelper.assertEquals(0, DateTimeHelper.getMillisecond(dateTime));
        AssertHelper.assertEquals(ZoneOffset.UTC, dateTime.getZone());
        julianDate = new JulianDate(2441683.5, TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        dateTime = julianDate.toDateTime(TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        AssertHelper.assertEquals(ZoneOffset.UTC, dateTime.getZone());
        julianDate = new JulianDate(2441683.5, TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        dateTime = julianDate.toDateTime(TimeStandard.COORDINATED_UNIVERSAL_TIME);
        AssertHelper.assertEquals(ZoneOffset.UTC, dateTime.getZone());
    }

    @Test
    public final void julianToGregorianDate() {
        JulianDate julianDate = new JulianDate(2451545.0, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        GregorianDate gregorianDate = julianDate.toGregorianDate();
        AssertHelper.assertEquals(2000, gregorianDate.getYear());
        AssertHelper.assertEquals(1, gregorianDate.getMonth());
        AssertHelper.assertEquals(1, gregorianDate.getDay());
        AssertHelper.assertEquals(12, gregorianDate.getHour());
        AssertHelper.assertEquals(0, gregorianDate.getMinute());
        AssertHelper.assertEquals(0, gregorianDate.getSecond());
        julianDate = new JulianDate(2453736.5, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        gregorianDate = julianDate.toGregorianDate();
        AssertHelper.assertEquals(2006, gregorianDate.getYear());
        AssertHelper.assertEquals(1, gregorianDate.getMonth());
        AssertHelper.assertEquals(1, gregorianDate.getDay());
        AssertHelper.assertEquals(0, gregorianDate.getHour());
        AssertHelper.assertEquals(0, gregorianDate.getMinute());
        AssertHelper.assertEquals(0, gregorianDate.getSecond());
        julianDate = new JulianDate(2441683.5, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        gregorianDate = julianDate.toGregorianDate();
        AssertHelper.assertEquals(1973, gregorianDate.getYear());
        AssertHelper.assertEquals(1, gregorianDate.getMonth());
        AssertHelper.assertEquals(1, gregorianDate.getDay());
        AssertHelper.assertEquals(0, gregorianDate.getHour());
        AssertHelper.assertEquals(0, gregorianDate.getMinute());
        AssertHelper.assertEquals(0, gregorianDate.getSecond());
        julianDate = new JulianDate(2441683.5, TimeStandard.COORDINATED_UNIVERSAL_TIME);
        GregorianDate internationalTimeGregorianDate = julianDate.toGregorianDate(TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        AssertHelper.assertNotEqual(gregorianDate, internationalTimeGregorianDate);
    }

    /**
    * Tests that an appropriate exception is thrown when attempting to construct a DateTime
    from a JulianDate prior to the earliest possible DateTime.
    */
    @Test
    public final void testJulianDateMinimumToDateTime() {
        final JulianDate date = JulianDate.getMinValue();
        AssertHelper.<ArgumentException> assertThrows(new TypeLiteral<ArgumentException>() {}, Action.of(() -> {
            ZonedDateTime unused = date.toDateTime();
        }));
    }

    /**
    * Tests conversion from a {@link ZonedDateTime} to a {@link JulianDate}.
    */
    @Test
    public final void dateTimeToJulian() {
        ZonedDateTime dateTime = DateTimeHelper.create(2000, 1, 1, 12, 0, 0);
        JulianDate julianDate = new JulianDate(dateTime);
        AssertHelper.assertEquals(2451545.0, julianDate.getTotalDays());
        dateTime = DateTimeHelper.create(2006, 1, 1, 0, 0, 0);
        julianDate = new JulianDate(dateTime);
        AssertHelper.assertEquals(2453736.5, julianDate.getTotalDays());
        dateTime = DateTimeHelper.create(1973, 1, 1, 0, 0, 0);
        julianDate = new JulianDate(dateTime);
        AssertHelper.assertEquals(2441683.5, julianDate.getTotalDays());
        ZonedDateTime localDateTime = DateTimeHelper.create(2000, 1, 1, 12, 0, 0, ZoneId.systemDefault());
        ZonedDateTime utc = DateTimeHelper.toUniversalTime(localDateTime);
        julianDate = new JulianDate(localDateTime);
        dateTime = julianDate.toDateTime();
        AssertHelper.assertEquals(utc, dateTime);
    }

    @Test
    public final void gregorianDateToJulian() {
        GregorianDate gregorianDate = new GregorianDate(2000, 1, 1, 12, 0, 0.0);
        JulianDate julianDate = new JulianDate(gregorianDate);
        AssertHelper.assertEquals(2451545.0, julianDate.getTotalDays());
        gregorianDate = new GregorianDate(2006, 1, 1, 0, 0, 0.0);
        julianDate = new JulianDate(gregorianDate);
        AssertHelper.assertEquals(2453736.5, julianDate.getTotalDays());
        gregorianDate = new GregorianDate(1973, 1, 1, 0, 0, 0.0);
        julianDate = new JulianDate(gregorianDate);
        AssertHelper.assertEquals(2441683.5, julianDate.getTotalDays());
        ZonedDateTime localDateTime = DateTimeHelper.create(2000, 1, 1, 12, 0, 0, ZoneId.systemDefault());
        ZonedDateTime utc = DateTimeHelper.toUniversalTime(localDateTime);
        julianDate = new JulianDate(localDateTime);
        gregorianDate = julianDate.toGregorianDate();
        AssertHelper.assertEquals(new GregorianDate(utc), gregorianDate);
    }

    /**
    * Tests conversion of a {@link ZonedDateTime} to a {@link JulianDate} when
    the {@link ZonedDateTime} is specified as a local time instead of as UTC.
    */
    @Test
    public final void testLocalDateTimeToJulian() {
        ZonedDateTime localNow = DateTimeHelper.now();
        ZonedDateTime utcNow = DateTimeHelper.toUniversalTime(localNow);
        JulianDate localJulianDate = new JulianDate(localNow);
        JulianDate utcJulianDate = new JulianDate(utcNow);
        Assert.assertTrue(localJulianDate.equalsEpsilon(utcJulianDate, Constants.Epsilon14));
        localNow = DateTimeHelper.now();
        utcNow = DateTimeHelper.toUniversalTime(localNow);
        localJulianDate = new JulianDate(localNow, TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        utcJulianDate = new JulianDate(utcNow, TimeStandard.INTERNATIONAL_ATOMIC_TIME);
        Assert.assertTrue(localJulianDate.equalsEpsilon(utcJulianDate, Constants.Epsilon14));
    }

    /**
    * Tests the "round-trip" conversion between a {@link ZonedDateTime} and a {@link JulianDate}.
    */
    @Test
    public final void julianDateTimeRoundTrip() {
        final ArrayList<ZonedDateTime> tempCollection$0 = new ArrayList<ZonedDateTime>();
        tempCollection$0.add(DateTimeHelper.utcNow());
        tempCollection$0.add(DateTimeHelper.create(2017, 8, 31, 13, 53, 32, 44, ZoneOffset.UTC));
        ArrayList<ZonedDateTime> dateTimes = tempCollection$0;
        // add a test date with all possible millisecond values
        for (int millisecond = 0; millisecond < 999; millisecond++) {
            dateTimes.add(DateTimeHelper.create(2017, 8, 31, 13, 53, 32, millisecond, ZoneOffset.UTC));
        }
        for (final ZonedDateTime dateTime : dateTimes) {
            JulianDate julianDate = new JulianDate(dateTime);
            ZonedDateTime roundTrip = julianDate.toDateTime();
            AssertHelper.assertEquals(dateTime.getYear(), roundTrip.getYear());
            AssertHelper.assertEquals(dateTime.getMonthValue(), roundTrip.getMonthValue());
            AssertHelper.assertEquals(dateTime.getDayOfMonth(), roundTrip.getDayOfMonth());
            AssertHelper.assertEquals(dateTime.getHour(), roundTrip.getHour());
            AssertHelper.assertEquals(dateTime.getMinute(), roundTrip.getMinute());
            AssertHelper.assertEquals(dateTime.getSecond(), roundTrip.getSecond());
            AssertHelper.assertEquals(DateTimeHelper.getMillisecond(dateTime), DateTimeHelper.getMillisecond(roundTrip));
        }
    }

    @Test
    public final void julianDateGregorianDateRoundTrip() {
        GregorianDate gregorianDate = new GregorianDate(2008, 1, 1, 1, 1, 1.00000000008);
        JulianDate julianDate = new JulianDate(gregorianDate);
        GregorianDate roundTrip = julianDate.toGregorianDate();
        AssertHelper.assertEquals(gregorianDate.getYear(), roundTrip.getYear());
        AssertHelper.assertEquals(gregorianDate.getMonth(), roundTrip.getMonth());
        AssertHelper.assertEquals(gregorianDate.getDay(), roundTrip.getDay());
        AssertHelper.assertEquals(gregorianDate.getHour(), roundTrip.getHour());
        AssertHelper.assertEquals(gregorianDate.getMinute(), roundTrip.getMinute());
        AssertHelper.assertEquals(gregorianDate.getSecond(), roundTrip.getSecond(), Constants.Epsilon13);
    }

    @Nonnull
    private static final TestContextRule rule$testContext = new TestContextRule();

    @Nonnull
    @Rule
    @ClassRule
    public static TestContextRule getRule$testContext() {
        return rule$testContext;
    }
}

File: DotNet/CesiumLanguageWriterTests/TestJulianDateConversions.cs
using System;
using System.Collections.Generic;
using CesiumLanguageWriter;
using NUnit.Framework;

namespace CesiumLanguageWriterTests
{
    /// <summary>
    /// A series of tests to exercise the various date and time types.
    /// </summary>
    [TestFixture]
    public class TestJulianDateConversions
    {
        /// <summary>
        /// Tests conversion from a <see cref="JulianDate"/> to a <see cref="DateTime"/>.
        /// </summary>
        [Test]
        public void JulianToDateTime()
        {
            JulianDate julianDate = new JulianDate(2451545.0, TimeStandard.CoordinatedUniversalTime);
            DateTime dateTime = julianDate.ToDateTime();
            Assert.AreEqual(2000, dateTime.Year);
            Assert.AreEqual(1, dateTime.Month);
            Assert.AreEqual(1, dateTime.Day);
            Assert.AreEqual(12, dateTime.Hour);
            Assert.AreEqual(0, dateTime.Minute);
            Assert.AreEqual(0, dateTime.Second);
            Assert.AreEqual(0, dateTime.Millisecond);
            Assert.AreEqual(DateTimeKind.Utc, dateTime.Kind);

            julianDate = new JulianDate(2453736.5, TimeStandard.CoordinatedUniversalTime);
            dateTime = julianDate.ToDateTime();
            Assert.AreEqual(2006, dateTime.Year);
            Assert.AreEqual(1, dateTime.Month);
            Assert.AreEqual(1, dateTime.Day);
            Assert.AreEqual(0, dateTime.Hour);
            Assert.AreEqual(0, dateTime.Minute);
            Assert.AreEqual(0, dateTime.Second);
            Assert.AreEqual(0, dateTime.Millisecond);
            Assert.AreEqual(DateTimeKind.Utc, dateTime.Kind);

            julianDate = new JulianDate(2441683.5, TimeStandard.CoordinatedUniversalTime);
            dateTime = julianDate.ToDateTime();
            Assert.AreEqual(1973, dateTime.Year);
            Assert.AreEqual(1, dateTime.Month);
            Assert.AreEqual(1, dateTime.Day);
            Assert.AreEqual(0, dateTime.Hour);
            Assert.AreEqual(0, dateTime.Minute);
            Assert.AreEqual(0, dateTime.Second);
            Assert.AreEqual(0, dateTime.Millisecond);
            Assert.AreEqual(DateTimeKind.Utc, dateTime.Kind);

            julianDate = new JulianDate(2441683.5, TimeStandard.InternationalAtomicTime);
            dateTime = julianDate.ToDateTime(TimeStandard.InternationalAtomicTime);
            Assert.AreEqual(DateTimeKind.Utc, dateTime.Kind);

            julianDate = new JulianDate(2441683.5, TimeStandard.InternationalAtomicTime);
            dateTime = julianDate.ToDateTime(TimeStandard.CoordinatedUniversalTime);
            Assert.AreEqual(DateTimeKind.Utc, dateTime.Kind);
        }

        [Test]
        public void JulianToGregorianDate()
        {
            JulianDate julianDate = new JulianDate(2451545.0, TimeStandard.CoordinatedUniversalTime);
            GregorianDate gregorianDate = julianDate.ToGregorianDate();
            Assert.AreEqual(2000, gregorianDate.Year);
            Assert.AreEqual(1, gregorianDate.Month);
            Assert.AreEqual(1, gregorianDate.Day);
            Assert.AreEqual(12, gregorianDate.Hour);
            Assert.AreEqual(0, gregorianDate.Minute);
            Assert.AreEqual(0, gregorianDate.Second);

            julianDate = new JulianDate(2453736.5, TimeStandard.CoordinatedUniversalTime);
            gregorianDate = julianDate.ToGregorianDate();
            Assert.AreEqual(2006, gregorianDate.Year);
            Assert.AreEqual(1, gregorianDate.Month);
            Assert.AreEqual(1, gregorianDate.Day);
            Assert.AreEqual(0, gregorianDate.Hour);
            Assert.AreEqual(0, gregorianDate.Minute);
            Assert.AreEqual(0, gregorianDate.Second);

            julianDate = new JulianDate(2441683.5, TimeStandard.CoordinatedUniversalTime);
            gregorianDate = julianDate.ToGregorianDate();
            Assert.AreEqual(1973, gregorianDate.Year);
            Assert.AreEqual(1, gregorianDate.Month);
            Assert.AreEqual(1, gregorianDate.Day);
            Assert.AreEqual(0, gregorianDate.Hour);
            Assert.AreEqual(0, gregorianDate.Minute);
            Assert.AreEqual(0, gregorianDate.Second);

            julianDate = new JulianDate(2441683.5, TimeStandard.CoordinatedUniversalTime);
            GregorianDate internationalTimeGregorianDate = julianDate.ToGregorianDate(TimeStandard.InternationalAtomicTime);
            Assert.AreNotEqual(gregorianDate, internationalTimeGregorianDate);
        }

        /// <summary>
        /// Test BUG40644 - JulianDate.ToDateTime rounds to the nearest millisecond
        /// </summary>
        [Test]
        [CSToJavaExclude("Java DateTime only supports millisecond precision")]
        public void TestBug40644()
        {
            JulianDate jd1 = new JulianDate(2451545, 0.0, TimeStandard.CoordinatedUniversalTime);
            JulianDate jd2 = new JulianDate(2451545, -0.0001, TimeStandard.CoordinatedUniversalTime);
            JulianDate jd3 = new JulianDate(2451545, 0.0001, TimeStandard.CoordinatedUniversalTime);

            DateTime date1 = jd1.ToDateTime();
            DateTime date2 = jd2.ToDateTime();
            DateTime date3 = jd3.ToDateTime();

            Assert.AreNotEqual(date1, date2);
            Assert.AreNotEqual(date1, date3);
            Assert.AreNotEqual(date2, date3);
        }

        /// <summary>
        /// Testing that ticks are accurately accounted for
        /// </summary>
        [Test]
        [CSToJavaExclude("Java DateTime only supports millisecond precision")]
        public void TestTicks()
        {
            DateTime yesterday = new DateTime(2008, 3, 4, 12, 5, 12, 24, DateTimeKind.Utc);
            DateTime someTimeLater = new DateTime(yesterday.Ticks + 50, DateTimeKind.Utc);

            JulianDate jDateYesterday = new JulianDate(yesterday);
            JulianDate jDateSomeTimeLater = new JulianDate(someTimeLater);

            Assert.AreNotEqual(jDateYesterday, jDateSomeTimeLater);
            Assert.AreEqual(jDateYesterday.Day, jDateSomeTimeLater.Day);
            Assert.AreEqual(jDateYesterday.SecondsOfDay, jDateSomeTimeLater.SecondsOfDay - 0.000005000, Constants.Epsilon9);
        }

        /// <summary>
        /// Tests that an appropriate exception is thrown when attempting to construct a DateTime
        /// from a JulianDate prior to the earliest possible DateTime.
        /// </summary>
        [Test]
        public void TestJulianDateMinimumToDateTime()
        {
            JulianDate date = JulianDate.MinValue;

            Assert.Throws<ArgumentOutOfRangeException>(() =>
            {
                DateTime unused = date.ToDateTime();
            });
        }

        /// <summary>
        /// Tests conversion from a <see cref="DateTime"/> to a <see cref="JulianDate"/>.
        /// </summary>
        [Test]
        public void DateTimeToJulian()
        {
            DateTime dateTime = new DateTime(2000, 1, 1, 12, 0, 0);
            JulianDate julianDate = new JulianDate(dateTime);
            Assert.AreEqual(2451545.0, julianDate.TotalDays);

            dateTime = new DateTime(2006, 1, 1, 0, 0, 0);
            julianDate = new JulianDate(dateTime);
            Assert.AreEqual(2453736.5, julianDate.TotalDays);

            dateTime = new DateTime(1973, 1, 1, 0, 0, 0);
            julianDate = new JulianDate(dateTime);
            Assert.AreEqual(2441683.5, julianDate.TotalDays);

            DateTime localDateTime = new DateTime(2000, 1, 1, 12, 0, 0, DateTimeKind.Local);
            DateTime utc = localDateTime.ToUniversalTime();
            julianDate = new JulianDate(localDateTime);
            dateTime = julianDate.ToDateTime();
            Assert.AreEqual(utc, dateTime);
        }

        [Test]
        public void GregorianDateToJulian()
        {
            GregorianDate gregorianDate = new GregorianDate(2000, 1, 1, 12, 0, 0);
            JulianDate julianDate = new JulianDate(gregorianDate);
            Assert.AreEqual(2451545.0, julianDate.TotalDays);

            gregorianDate = new GregorianDate(2006, 1, 1, 0, 0, 0);
            julianDate = new JulianDate(gregorianDate);
            Assert.AreEqual(2453736.5, julianDate.TotalDays);

            gregorianDate = new GregorianDate(1973, 1, 1, 0, 0, 0);
            julianDate = new JulianDate(gregorianDate);
            Assert.AreEqual(2441683.5, julianDate.TotalDays);

            DateTime localDateTime = new DateTime(2000, 1, 1, 12, 0, 0, DateTimeKind.Local);
            DateTime utc = localDateTime.ToUniversalTime();
            julianDate = new JulianDate(localDateTime);
            gregorianDate = julianDate.ToGregorianDate();
            Assert.AreEqual(new GregorianDate(utc), gregorianDate);
        }

        /// <summary>
        /// Tests conversion of a <see cref="DateTime"/> to a <see cref="JulianDate"/> when
        /// the <see cref="DateTime"/> is specified as a local time instead of as UTC.
        /// </summary>
        [Test]
        public void TestLocalDateTimeToJulian()
        {
            DateTime localNow = DateTime.Now;
            DateTime utcNow = localNow.ToUniversalTime();

            JulianDate localJulianDate = new JulianDate(localNow);
            JulianDate utcJulianDate = new JulianDate(utcNow);

            Assert.IsTrue(localJulianDate.EqualsEpsilon(utcJulianDate, Constants.Epsilon14));

            localNow = DateTime.Now;
            utcNow = localNow.ToUniversalTime();

            localJulianDate = new JulianDate(localNow, TimeStandard.InternationalAtomicTime);
            utcJulianDate = new JulianDate(utcNow, TimeStandard.InternationalAtomicTime);

            Assert.IsTrue(localJulianDate.EqualsEpsilon(utcJulianDate, Constants.Epsilon14));
        }

        /// <summary>
        /// Tests the "round-trip" conversion between a <see cref="DateTime"/> and a <see cref="JulianDate"/>.
        /// </summary>
        [Test]
        public void JulianDateTimeRoundTrip()
        {
            var dateTimes = new List<DateTime>
            {
                DateTime.UtcNow,
                // a previous version of the code didn't round-trip for this particular date due to 
                // an error when values were rounded
                new DateTime(2017, 8, 31, 13, 53, 32, 44, DateTimeKind.Utc)
            };

            // add a test date with all possible millisecond values
            for (int millisecond = 0; millisecond < 999; millisecond++)
            {
                dateTimes.Add(new DateTime(2017, 8, 31, 13, 53, 32, millisecond, DateTimeKind.Utc));
            }

            foreach (DateTime dateTime in dateTimes)
            {
                JulianDate julianDate = new JulianDate(dateTime);
                DateTime roundTrip = julianDate.ToDateTime();
                Assert.AreEqual(dateTime.Year, roundTrip.Year);
                Assert.AreEqual(dateTime.Month, roundTrip.Month);
                Assert.AreEqual(dateTime.Day, roundTrip.Day);
                Assert.AreEqual(dateTime.Hour, roundTrip.Hour);
                Assert.AreEqual(dateTime.Minute, roundTrip.Minute);
                Assert.AreEqual(dateTime.Second, roundTrip.Second);
                Assert.AreEqual(dateTime.Millisecond, roundTrip.Millisecond);
            }
        }

        [Test]
        public void JulianDateGregorianDateRoundTrip()
        {
            GregorianDate gregorianDate = new GregorianDate(2008, 1, 1, 1, 1, 1.00000000008);
            JulianDate julianDate = new JulianDate(gregorianDate);
            GregorianDate roundTrip = julianDate.ToGregorianDate();
            Assert.AreEqual(gregorianDate.Year, roundTrip.Year);
            Assert.AreEqual(gregorianDate.Month, roundTrip.Month);
            Assert.AreEqual(gregorianDate.Day, roundTrip.Day);
            Assert.AreEqual(gregorianDate.Hour, roundTrip.Hour);
            Assert.AreEqual(gregorianDate.Minute, roundTrip.Minute);
            Assert.AreEqual(gregorianDate.Second, roundTrip.Second, Constants.Epsilon13);
        }
    }
}